beacon-chain/rpc/validator/proposer.go

```diff
@@ -497,10 +498,56 @@ func (vs *Server) depositTrie(ctx context.Context, canonicalEth1DataHeight *big.
 		depositTrie.Insert(depHash[:], int(insertIndex))
 		insertIndex++
 	}
+	valid, err := vs.validateDepositTrie(depositTrie, canonicalEth1Data)
+	// Log a warning here, as the cached trie is invalid is invalid.
+	if !valid {
+		log.Warnf("Cached deposit trie is invalid, rebuilding it now: %v", err)
+		return vs.rebuildDepositTrie(ctx, canonicalEth1Data, canonicalEth1DataHeight)
+	}
+
+	return depositTrie, nil
+}
+
+// rebuilds our deposit trie by recreating it from all processed deposits till
+// specified eth1 block height.
+func (vs *Server) rebuildDepositTrie(ctx context.Context, canonicalEth1Data *ethpb.Eth1Data, canonicalEth1DataHeight *big.Int) (*trieutil.SparseMerkleTrie, error) {
+	ctx, span := trace.StartSpan(ctx, "ProposerServer.rebuildDepositTrie")
+	defer span.End()
+
+	deposits := vs.DepositFetcher.AllDeposits(ctx, canonicalEth1DataHeight)
+	trieItems := make([][]byte, 0, len(deposits))
+	for _, dep := range deposits {
+		depHash, err := dep.Data.HashTreeRoot()
+		if err != nil {
+			return nil, errors.Wrap(err, "could not hash deposit data")
+		}
+		trieItems = append(trieItems, depHash[:])
+	}
+	depositTrie, err := trieutil.GenerateTrieFromItems(trieItems, params.BeaconConfig().DepositContractTreeDepth)
+	if err != nil {
+		return nil, err
+	}
 
+	valid, err := vs.validateDepositTrie(depositTrie, canonicalEth1Data)
+	// Log an error here, as even with rebuilding the trie, it is still invalid.
+	if !valid {
+		log.Errorf("Rebuilt deposit trie is invalid: %v", err)
```

Comment:
Shouldn't `err` be returned from the function instead of just logging? We essentially have a problem with generating the deposit trie.

Reply:
The purpose is to log more strongly here, even if we fail creating a valid deposit trie we can still propose a block( as long as there are no deposits)